packages/runtime/runtime-definitions/src/id-compressor/idCompressor.ts

```diff
@@ -0,0 +1,140 @@
+	 * @param range - the range of session-local IDs to finalize.
+	 */
+	finalizeCreationRange(range: IdCreationRange): void;
+
+	/**
+	 * Returns a range of local IDs created by this session in a format for sending to the server for finalizing.
+	 * The range will include all local IDs generated via calls to `generateCompressedId` since the last time this method was called.
+	 * @returns the range of session-local IDs, which may be empty. This range must be sent to the server for ordering before
+	 * it is finalized. Ranges must be sent to the server in the order that they are taken via calls to this method.
+	 */
+	takeNextCreationRange(): IdCreationRange;
+
+	/**
+	 * Returns an iterable of all IDs created by this compressor.
+	 */
+	getAllIdsFromLocalSession(): IterableIterator<SessionSpaceCompressedId>;
+
+	/**
+	 * Returns a persistable form of the current state of this `IdCompressor` which can be rehydrated via `IdCompressor.deserialize()`.
+	 * This includes finalized state as well as un-finalized state and is therefore suitable for use in offline scenarios.
+	 */
+	serialize(
+		withSession: boolean,
+	): SerializedIdCompressorWithOngoingSession | SerializedIdCompressorWithNoSession;
+
+	/**
+	 * Returns a persistable form of the current state of this `IdCompressor` which can be rehydrated via `IdCompressor.deserialize()`.
+	 * This includes finalized state as well as un-finalized state and is therefore suitable for use in offline scenarios.
+	 */
+	serialize(withSession: true): SerializedIdCompressorWithOngoingSession;
+
+	/**
+	 * Returns a persistable form of the current state of this `IdCompressor` which can be rehydrated via `IdCompressor.deserialize()`.
+	 * This only includes finalized state and is therefore suitable for use in summaries.
+	 */
+	serialize(withSession: false): SerializedIdCompressorWithNoSession;
+	serialize(withSession: boolean): SerializedIdCompressor;
+}
+
+/**
+ * This is the set of operations on an IdCompressor that DDS author
+ * would leverage. These functions allow for the generation of Ids and
+ * translation between different Id spaces.
+ */
+export interface IIdCompressor {
+	localSessionId: SessionId;
+
+	/**
+	 * Generates a new compressed ID or returns an existing one.
+	 * This should ONLY be called to generate IDs for local operations.
+	 * @param override - Specifies a specific string to be associated with the returned compressed ID.
+	 * Performance note: assigning override strings incurs a performance overhead.
+	 * @returns an existing ID if one already exists for `override`, and a new local ID otherwise. The returned ID is in session space.
+	 */
+	generateCompressedId(): SessionSpaceCompressedId;
+	/**
+	 * Normalizes a session space ID into op space.
+	 * @param id - the local ID to normalize.
+	 * @returns the ID in op space.
+	 */
+	normalizeToOpSpace(id: SessionSpaceCompressedId): OpSpaceCompressedId;
```

Comment:
Seeing the normalization methods on this interface made me think two things:
1. We should explore the possibility of having a special field (e.g., "__fluid_id") that gets automatically normalized if used in persisted state.
2. Regardless of 1, we should have really good docs on these methods explaining normalization and why an author needs to use it.